src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1347 @@
+  }
+  else if(type.id()==ID_struct)
+  {
+    irep_idt tag=to_struct_type(type).get_tag();
+    return (tag==irep_idt("java.lang.String"));
+  }
+  else
+    return false;
+}
+
+static bool is_java_string_pointer_type(const typet& type)
+{
+  if(type.id()==ID_pointer)
+  {
+    pointer_typet pt=to_pointer_type(type);
+    typet subtype=pt.subtype();
+    return is_java_string_type(subtype);
+  }
+  else return false;
+}
+
+/*******************************************************************    \
+
+Function: string_refine_preprocesst::replace_string_pointer
+
+  Inputs: an expression
+
+ Outputs: an expression
+
+ Purpose: replace string pointers by expressions as expected by the string refinement;
+          should only be used in arguments of __CPROVER_uninterpreted functions.
+
+\*******************************************************************/
+
+exprt string_refine_preprocesst::replace_string_pointer(const exprt& in)
+{
+  if(is_java_string_pointer_type(in.type()))
+  {
+    exprt clsid=
+      symbol_exprt("java::java.lang.String", string_typet());
+    // TODO: we should use a function which is not currently in the code base
+    // to get the class identifier in this way:
+    // get_class_identifier_field(in, symbol_typet("java::java.lang.Object"), ns);
+    auto it=hidden_strings.find(in.op0());
+    if(in.id()==ID_typecast && it!=hidden_strings.end())
+	return it->second;
+    else
+      return string_exprt(in, clsid, true);
+  }
+  else return in;
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_cprover_string_assign
+
+  Inputs: a goto_program, a position in this program, an expression and a
+          location
+
+ Outputs: an expression
+
+ Purpose: Introduce a temporary variable for cprover strings;
+          returns the cprover_string corresponding to rhs if it is a string
+          pointer and the original rhs otherwise.
+
+\*******************************************************************/
+
+exprt string_refine_preprocesst::make_cprover_string_assign(
+  goto_programt & goto_program,
+  goto_programt::instructionst::iterator & i_it,
+  const exprt & rhs,
+  source_locationt location)
+{
+  if(is_java_string_pointer_type(rhs.type()))
+  {
+    auto it=java_to_cprover_strings.find(rhs);
+    if(it!=java_to_cprover_strings.end())
+      return it->second;
+    else
+    {
+      debug() << "string_refine_preprocesst::make_cprover_string_assign "
+              << rhs.id() << eom;
+      exprt clsid=symbol_exprt("java::java.lang.String",string_typet());
+      // get_class_identifier_field(rhs, symbol_typet("java::java.lang.Object"), ns);
+      string_exprt new_rhs(rhs, clsid, true);
+      symbol_exprt lhs=new_symbol("cprover_string", new_rhs.type());
+      symbol_exprt array_lhs = new_symbol(
+        "cprover_string_array", new_rhs.content().type());
+      exprt content=new_rhs.content();
+      // This does not look nice:
```

Comment:
Using `op1` to access the content of the string was too dependent of the implementation of `string_exprt` so instead we now use the `content()` function: https://github.com/diffblue/cbmc/pull/378/commits/b8afe554efc74265bb0cdfa1eee6055dd87c1b8f